src/expression.c

```diff
@@ -6104,8 +6104,9 @@ Expression *TypeidExp::semantic(Scope *sc)
     {
         /* Get the dynamic type, which is .classinfo
          */
-        e = new DotIdExp(ea->loc, ea, Id::classinfo);
-        e = e->semantic(sc);
+        ea = ea->semantic(sc);
+        e = new TypeidExp(ea->loc, ea);
+        e->type = Type::typeinfoclass->type;
```

Comment:
Thanks!  It's a little worrying how sparse the test coverage is... I'm still not really sure why those last two asserts in the test case fail, but at least they work the same before and after this pull request.